hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTask.java

```diff
@@ -0,0 +1,304 @@
+              deleteOmDirectoryInfoOnNamespaceDB(oldDirectoryInfo);
+            } else {
+              LOG.warn("Update event does not have the old dirInfo for {}.",
+                      updatedKey);
+            }
+            writeOmDirectoryInfoOnNamespaceDB(updatedDirectoryInfo);
+            break;
+
+          default:
+            LOG.debug("Skipping DB update event : {}",
+                    omdbUpdateEvent.getAction());
+          }
+        }
+      } catch (IOException ioEx) {
+        LOG.error("Unable to process Namespace Summary data in Recon DB. ",
+                ioEx);
+        return new ImmutablePair<>(getTaskName(), false);
+      }
+    }
+    LOG.info("Completed a process run of NSSummaryTask");
+    return new ImmutablePair<>(getTaskName(), true);
+  }
+
+  @Override
+  public Pair<String, Boolean> reprocess(OMMetadataManager omMetadataManager) {
+    // actually fileTable with FSO
+    Table keyTable = omMetadataManager.getKeyTable();
+
+    TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+            keyTableIter = keyTable.iterator();
+
+    try {
+      // reinit Recon RocksDB's namespace CF.
+      reconNamespaceSummaryManager.initNSSummaryTable();
+
+      while (keyTableIter.hasNext()) {
+        Table.KeyValue<String, OmKeyInfo> kv = keyTableIter.next();
+        OmKeyInfo keyInfo = kv.getValue();
+        writeOmKeyInfoOnNamespaceDB(keyInfo);
+      }
+
+      Table dirTable = omMetadataManager.getDirectoryTable();
+      TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+              dirTableIter = dirTable.iterator();
+
+      while (dirTableIter.hasNext()) {
+        Table.KeyValue<String, OmDirectoryInfo> kv = dirTableIter.next();
+        OmDirectoryInfo directoryInfo = kv.getValue();
+        writeOmDirectoryInfoOnNamespaceDB(directoryInfo);
+      }
+
+    } catch (IOException ioEx) {
+      LOG.error("Unable to reprocess Namespace Summary data in Recon DB. ",
+              ioEx);
+      return new ImmutablePair<>(getTaskName(), false);
+    }
+
+    LOG.info("Completed a reprocess run of NSSummaryTask");
+    return new ImmutablePair<>(getTaskName(), true);
+  }
+
+  private void writeOmKeyInfoOnNamespaceDB(OmKeyInfo keyInfo)
+          throws IOException {
+    long parentObjectId = keyInfo.getParentObjectID();
+    NSSummary nsSummary = reconNamespaceSummaryManager
+            .getNSSummary(parentObjectId);
+    if (nsSummary == null) {
+      nsSummary = new NSSummary();
+    }
+    int numOfFile = nsSummary.getNumOfFiles();
+    long sizeOfFile = nsSummary.getSizeOfFiles();
+    int[] fileBucket = nsSummary.getFileSizeBucket();
+    nsSummary.setNumOfFiles(numOfFile + 1);
+    long dataSize = keyInfo.getDataSize();
+    nsSummary.setSizeOfFiles(sizeOfFile + dataSize);
+    int binIndex = ReconUtils.getBinIndex(dataSize);
+
+    // make sure the file is within our scope of tracking.
+    if (binIndex >= 0 && binIndex < ReconConstants.NUM_OF_BINS) {
```

Comment:
The FileSizeCountTask was supposed to map all files >1PB to the last bucket. Hence, there is no overflow of bin index conceptually. Is that not the case?

Reply:
I have a separate function to compute bin index, but yes, I can move this check to the function.